106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+
+If, during an ongoing auto-rebalancing, the `KafkaRebalance` custom resource is not there anymore on the next reconciliation, it could mean the user deleted it while the operator was stopped/crashed/not running.
+In this case, the FSM will assume it as `NotReady` so falling in the last case above.
+
+## Affected/not affected projects
+
+This change will affect the Strimzi cluster operator and a new repository named `anomaly-detector-notifier` will be added under the Strimzi organisation.
+
+## Rejected Alternatives
+
+### Using custom resource for interaction between Operator and Cruise Control I
+
+This alternative is about using a Kubernetes custom resource to create a two-way interaction between the operator and Cruise Control.
+The idea was to create a Kubernetes custom resource named `KafkaAnomaly` everytime an anomaly was detected.
+The user will get updates regarding the anomaly fix through the generated `KafkaAnomaly` resource which would be updated by the operator by requesting the `state` endpoint with `anomaly_detector` and `executor` substates.
+
+Pros:
+* More hold on the self-healing process since everything is driven using the Kafka custom resource.
+
+Cons:
+* Very tight coupling with the operator.
+* Would be hard to manage multiple `KafkaAnomaly` custom resources (For example, deletion when anomaly is fixed etc.)
+
+### Using custom resource for interaction between Operator and Cruise Control II
+
+This alternative is similar to alternative 1 where we will use a Kubernetes custom resource to create a two-way interaction between the operator and Cruise Control.
+The idea was to create a Kubernetes custom resource named `KafkaAnomaly` everytime an anomaly was detected and both the operator and the Notifier would watch the resource for updates.
+But with this approach the operator will be responsible to make decision regarding the anomaly should be fixed or not
+
+Pros:
+* Allows the operator to ignore the anomalies is some task is already running in the cluster like rolling, rebalance etc.
+
+Cons:
+* Very tight coupling with the operator.
+* Delaying the anomaly detector progress.
+
+### Using Kubernetes Events but Cruise Control controls auto-rebalance
+
+This alternative is to let Cruise Control handle the self-healing.
+Whenever an anomaly is detected by Cruise Control, our notifier will generate an event to alert the operator that an anomaly was detected in the cluster
+But the fix would be run by Cruise Control itself and not the operator
+
+Pros:
+* Loose coupling with the operator
+* Faster decison making as Cruise Control runs the fix
+
+Cons:
+* Operator wouldn't play any role in the process
+
+## Future Scope
+In the future, we plan to introduce auto-rebalance for topic and metrics related imbalances.
+AS this feature evolves we can even think of having ways to fix issue like disk failure and broker failures automatically since the fix would be driven by the operator.
```

Comment:
```suggestion
As this feature evolves we can even think of having ways to fix issues like disk failure and broker failures automatically since the fix would be driven by the operator.
```